src/platforms/esp32/components/avm_builtins/i2c_resource.c

```diff
@@ -0,0 +1,743 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later
+ */
+
+#include <sdkconfig.h>
+#ifdef CONFIG_AVM_ENABLE_I2C_RESOURCE_NIFS
+
+#include <string.h>
+
+// Note: The i2c.h legacy driver will be deprecated in 5.2
+// TODO: Migrate to i2c_master.h, once we drop support for IDF SDK 4.x
+#include <driver/i2c.h>
+#include <esp_log.h>
+
+#include <context.h>
+#include <defaultatoms.h>
+#include <erl_nif_priv.h>
+#include <globalcontext.h>
+#include <interop.h>
+#include <module.h>
+#include <nifs.h>
+#include <term.h>
+
+// #define ENABLE_TRACE
+#include <trace.h>
+
+#include <esp32_sys.h>
+#include <sys.h>
+
+#define TAG "i2c_resource"
+
+#define CHECK_ERROR(ctx, err, msg)                                                      \
+if (UNLIKELY(err != ESP_OK)) {                                                          \
+    ESP_LOGE(TAG, msg ": err: %i.", err);                                               \
+    if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {             \
+        return OUT_OF_MEMORY_ATOM;                                                      \
+    }                                                                                   \
+    return create_error_tuple(ctx, esp_err_to_term(ctx->global, err));                  \
+}
+
+static ErlNifResourceType *i2c_resource_type;
+
+#define EINPROGRESS_ATOMSTR (ATOM_STR("\xB", "einprogress"))
+#define I2C_ATOMSTR (ATOM_STR("\x4", "$i2c"))
+
+struct I2CResource
+{
+    term transmitting_pid;
+    i2c_port_t i2c_num;
+    i2c_cmd_handle_t cmd;
+};
+
+static term create_pair(Context *ctx, term term1, term term2)
+{
+    term ret = term_alloc_tuple(2, &ctx->heap);
+    term_put_tuple_element(ret, 0, term1);
+    term_put_tuple_element(ret, 1, term2);
+
+    return ret;
+}
+
+static term create_error_tuple(Context *ctx, term reason)
+{
+    return create_pair(ctx, ERROR_ATOM, reason);
+}
+
+static bool is_i2c_resource(GlobalContext *global, term t)
+{
+    bool ret = term_is_tuple(t)
+        && term_get_tuple_arity(t) == 3
+        && globalcontext_is_term_equal_to_atom_string(global, term_get_tuple_element(t, 0), I2C_ATOMSTR)
+        && term_is_binary(term_get_tuple_element(t, 1))
+        && term_is_reference(term_get_tuple_element(t, 2));
+
+    return ret;
+}
+
+static bool to_i2c_resource(term i2c_resource, struct I2CResource **rsrc_obj, Context *ctx)
+{
+    if (!is_i2c_resource(ctx->global, i2c_resource)) {
+        return false;
+    }
+    void *rsrc_obj_ptr;
+    if (UNLIKELY(!enif_get_resource(erl_nif_env_from_context(ctx), term_get_tuple_element(i2c_resource, 1), i2c_resource_type, &rsrc_obj_ptr))) {
+        return false;
+    }
+    *rsrc_obj = (struct I2CResource *) rsrc_obj_ptr;
+
+    return true;
+}
+
+//
+// i2c:open_nif/1
+//
+
+static term nif_open(Context *ctx, int argc, term argv[])
```

Comment:
What is the confusion here, that it would fail in the future, for example, if the symbol were defined elsewhere?   This is a static function, so it is not exported outside of the compilation unit.

Reply:
I have added `i2c_` to help avoid name conflicts in the future.